File: ruby/scrabble-score/scrabble.rb
class Scrabble

  VALUES = {
    A: 1, B: 3, C: 3, D: 2, E: 1, F: 4, G: 2, H: 4, I: 1, J: 8, K: 5, L: 1, M: 3,
    N: 1, O: 1, P: 3, Q:10, R: 1, S: 1, T: 1, U: 1, V: 4, W: 4, X: 8, Y: 4, Z:10
  }

  def initialize(word)
    @word = word.to_s.strip.upcase
  end

  def score
    @word.each_char.reduce(0) { |memo, obj| memo += VALUES[obj.to_sym] }
  end

  def self.score(word)
    self.new(word).score
  end

end

File: elixir/scrabble-score/scrabble.exs
defmodule Scrabble do

  @doc """
  Calculate the scrabble score for the word.
  """
  @spec score(String.t) :: non_neg_integer
  def score(word) do
    word
    |> String.upcase
    |> String.graphemes
    |> Enum.reduce(0, &(&2 + score_for(&1)))
  end

  @spec score_for(String.t) :: non_neg_integer
  defp score_for(letter) when letter in ~w(A    E   I  L N O  R S T U      ), do: 1
  defp score_for(letter) when letter in ~w(    D  G                        ), do: 2
  defp score_for(letter) when letter in ~w( B C         M   P              ), do: 3
  defp score_for(letter) when letter in ~w(      F H                 V W Y ), do: 4
  defp score_for(letter) when letter in ~w(           K                    ), do: 5
  defp score_for(letter) when letter in ~w(          J                  X  ), do: 8
  defp score_for(letter) when letter in ~w(                  Q            Z), do: 10
  defp score_for(_), do: 0

end
